teps/0147-tekton-artifacts-phase1.md

```diff
@@ -0,0 +1,285 @@
+```yaml
+taskSpec:
+  steps:
+    - name: artifact-reporting
+      image: bash:latest
+      script: |
+        #!/usr/bin/env bash
+        # Download some artifacts
+        # Upload some artifacts
+        echo -n "{\"inputs\":[{\"name\":\"artifact1\",\"values\":[{\"url\":\"some-uri\",\"digest\":\"some-digest\"}]}], \"outputs\":[{\"name\":\"artifact2\",\"values\":[{\"url\":\"some-uri\",\"digest\":\"some-digest\"}]}]}" | tee /tekton/steps/artifact-reporting/artifacts/provenance.json
+```
+
+#### Name
+Name is a `Task`/`Step` defined name. It is useful for referencing (phase 2) and UI. The name the artifact under a particular category (`input` or `output`) must be unique.
+
+#### Values
+Values is a list of objects with properties: [`uri`](#uri) and [`digest`](#digest).
+
+##### `uri`
+
+The `uri` component identifies an artifact within the artifact storage in use.
+The storage itself is included as part of the `TaskRun` and `PipelineRun` definition.
+
+The recommended (we could choose to enforce this) format for the`uri` is [purl](https://github.com/package-url/purl-spec/blob/master/PURL-TYPES.rst). Example values:
+
+
+- `pkg:generic/repo?download_url=/path/to/artifact`
+    - This path would also need to be accessible by downstream Tasks for this to work. The user of the `Task/StepAction` would need to orchestrate it carefully.
+- `pkg:generic/test-results?download_url=https://localobjectstore/filename.tar.gz`
+    - This approach would work for sharing artifacts (files/tarred-folders) between Tasks without the need for a `Persistent Volume`.
+- `pkg:github/package-url/purl-spec@244fd47e07d1004`
+- `pkg:docker/cassandra@latest`
+- `pkg:huggingface/distilbert-base-uncased@043235d6088ecd3dd5fb5ca3592b6913fd516027` (ML models)
+- `pkg:maven/org.apache.xmlgraphics/batik-anim@1.9.1`
+
+`purl-spec` also provides a common standard that `Tasks/Steps` can leverage so that output metadata information of one `Task/Step` could be consumed by another in the future (targetting phase 2). Additionally, it becomes a standard way to distinguish between types of artifacts. [package-url](https://github.com/package-url) also provides client libraries in multiple languages to parse the `purl`. This is something users could leverage.
+
+##### `digest`
+
+The `digest` component includes the result of applying an hash function to the artifact itself.
+The `digest` is in the format `hash-function:<digest>`. Example values:
+
+- `md5:471a3a9ddc1d6e699190b9f4ca88aadf`
+- `sha256:8796357729cfd877cf8fa7d45a8ab3524d9249c23a0bf68bb0026c0783b881d2`
+
+### From Step into Status
+
+The artifact provenance data (i.e. metadata content) is extracted from the local storage using the same mechanism used for results. Results and artifacts data is collated together and stored either in the termination message or in a sidecar log, depending on configuration.
```

Comment:
Uh oh. Neither termination message nor sidecar log seem like a good idea for storing an unbounded list of items.

Reply:
As a simple experiment, I dumped 7000 artifacts with metadata  {'uri': 'https://github.com/tektoncd/pipeline.git', 'digest': 'sha256:6042500cf4b44023ea1894effe7890666b0c5c7871ed83a97c36c76ae560bb9b'} to a file. The file was about 1MB in size. That's doable via sidecar logs with a feature flag toggle `max-result-size`: 4096 (default).

But in the long run, yes, we need a better storage solution. We have some options but we can explore them in the future after artifact support is out.